src/libraries/System.Linq.AsyncEnumerable/src/System/Linq/Sequence.cs

```diff
@@ -0,0 +1,130 @@
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="step"/> is zero and <paramref name="endInclusive"/> does not equal <paramref name="start"/>.</exception>
+        public static IAsyncEnumerable<T> Sequence<T>(T start, T endInclusive, T step) where T : INumber<T>
+        {
+            if (start is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(start));
+            }
+
+            if (endInclusive is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(endInclusive));
+            }
+
+            if (step is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(step));
+            }
+
+            if (step > T.Zero)
+            {
+                // Presumed to be the most common case, step > 0. Validate that endInclusive >= start, as otherwise we can't easily
+                // guarantee that the sequence will terminate.
+                if (endInclusive < start)
+                {
+                    ThrowHelper.ThrowArgumentOutOfRangeException(nameof(endInclusive));
+                }
+
+                // Otherwise, just produce an incrementing sequence.
+                return Incrementing(start, endInclusive, step);
+            }
+            else if (step < T.Zero)
+            {
+                // step < 0. Validate that endInclusive <= start, as otherwise we can't easily guarantee that the sequence will terminate.
+                if (endInclusive > start)
+                {
+                    ThrowHelper.ThrowArgumentOutOfRangeException(nameof(endInclusive));
+                }
+
+                // Then produce the decrementing sequence.
+                return Decrementing(start, endInclusive, step);
+            }
+            else
+            {
+                // step == 0. If start != endInclusive, then the sequence would be infinite. As such, we validate
+                // that they're equal, and if they are, we return a sequence that yields the start/endInclusive value once.
+                if (start != endInclusive)
+                {
+                    ThrowHelper.ThrowArgumentOutOfRangeException(nameof(step));
+                }
+
+                return Repeat(start, 1);
+            }
+
+            static async IAsyncEnumerable<T> Incrementing(T start, T endInclusive, T step)
+            {
+                Debug.Assert(step > T.Zero);
+
+                yield return start;
+
+                while (true)
+                {
+                    T next = start + step;
+
+                    if (next >= endInclusive || next < start)
+                    {
+                        if (next == endInclusive && start != next)
+                        {
+                            yield return next;
+                        }
+
+                        yield break;
+                    }
+
+                    yield return next;
```

Comment:
Tanner had opinions about when a floating point hits an additive loss situation.  I feel like it was that here we should have `if (next == start) { yield break; }` here.

Otherwise a big number (1M?) plus a small number (Epsilon) will turn into Repeat(next, infinite)

Reply:
Y....ep.  Then it becomes "if next is last, or next is too big, or next overflowed, or next didn't make progress".  Which sounds like all of the terminating conditions.

Reply:
`next <= start` can't detect some overflow cases such as `Enumerable.Sequence<byte>(0, 255, 2)`. A better implementation to detect overflows would be checking if `next < current`.